include/linux/skbuff.h

```diff
@@ -731,12 +731,10 @@ struct sk_buff {
 				 */
 				unsigned long		dev_scratch;
 #ifdef CONFIG_SECURITY_TEMPESTA
-                                struct {
-                                        __u8    present : 1;
-                                        __u8    tls_type : 7;
-                                        __u16   flags : 16;
```

Comment:
Still need `flags` to notify when skb is transmitted, e.g. https://github.com/tempesta-tech/tempesta/pull/2108.
Of course, `on_tcp_entail` can do this as well, but it is not currently used for this purpose, at least not for all control frames.

Reply:
We should implement additional  on_send_ function in tempesta for 2108, not use flags in skb. Moreover we remove additional callback from kernel code, so 2108 should be reworked

Reply:
No, it can't be in `on_send` callback, it only indicates entry into the write queue of sk, but the counter decrement must happen at the time of actual transmission, because only at that point in time the skb is moved to the TCP protocol transmission process, but not accumulated on the tempestat side, i.e. `tcp_write_xmit` or its callers, e.g. `__tcp_push_pending_frames`. This is exactly the meaning of this counter, to avoid control frame flooding.

Reply:
`ss_skb_on_tcp_entail` is called before pushing skb to socket write queue. It is a good place to decrement counter.

Reply:
Applications which works in user space and counted control frames don't know nothing about `__tcp_push_pending_frames` and there is no problem.

Reply:
> Applications which works in user space and counted control frames don't know nothing about `__tcp_push_pending_frames` and there is no problem.

I don't get your point.

Reply:
> ss_skb_on_tcp_entail is called before pushing skb to socket write queue. It is a good place to decrement counter.

Here's a little subtle and interesting. It's enqueued for writing, that's true, but that's not the whole truth, it's enqueued when TCP is ready to send more data, because you changed the code to put it in `tcp_push_pending_frames`, right?

Reply:
> May be we should limit count of all frame types when we push to socket write queue

Just control frame types are enough, so I think `tfw_cb.flag` is always needed, no matter where you decrement the counter.

Reply:
We don't limit count of all frame types, may be we should do it and do not push control frames, if we have no enough send window.  We should save them in H2Ctx in special queue. If we implement it, we will decrement count of control frames in `ss_skb_on_tcp_entail` when tcp is ready to send.

Reply:
However, if you follow your approach, control frames will be reordered with other frames from different streams. (Note that control frames do not belong to any stream and may be interleaved with other frames in the skb queue.)

Reply:
@EvgeniiMekhanik could you please comment in #2108 how do you see which `on_send_` function(s) must be implemented and where they should be called from according to the new design in #1973?

Reply:
I'd like to correct my above statement, now I think it is inappropriate to count frames in `ss_skb_on_tcp_entail`, because it is only counted when a frame/skb is actually transmitted (when both congestion window and sender window allow), i.e. inside `tcp_tfw_sk_write_xmit`, which happens before `tcp_transmit_skb`, which in turn transmits the skb to lower layers (or retransmits it later).

Reply:
I think we can postpone such frames if it is really necessary (and send them only when we have enough window) and count them in `ss_skb_on_tcp_entail`. I think that we move tls_encrypt to Tempesta code and there is no sense to have one more callback in linux kernel only to count control frames. User space applications like nginx doesn't know anything about `tcp_transmit_skb`. How they can count control frames?

Reply:
> User space applications like nginx doesn't know anything about tcp_transmit_skb. How they can count control frames?

@EvgeniiMekhanik @krizhanovsky @const-t 

_I have to update some thoughts._

No, the user program relies on sockbuf to determine whether the skb send is in-flight. If sockbuf is full, the `send()` system call will fail because the peer does not provide enough window, or in non-blocking fd, the write event is shelved, then the counter will accumulate, and then the user program will know that it is in a flood attack.

However, in our case, it's a completely different story. tempesta does not use sockbuf, but directly enqueues to sk's sk_write_queue, so it will have memory risks unless we know that skb has started to be transmitted, that is, in-flight. How to know this fact? [The only place is that `tcp_transmit_skb` successfully handles skb](https://github.com/tempesta-tech/linux-5.10.35-tfw/blob/d12a8e0d13890204c4ae64051fe3844dd4b40ecf/net/ipv4/tcp_output.c#L2829C1-L2833C51). So this is why I suggest that we should apply the counting logic there. Neither on_send (called in ss_tx_action) nor entail (corresponding to the original prepare_xmit), or even write_xmit callback, can meet our needs to mitigate the flood of control frame attacks.

And the method suggested by @EvgeniiMekhanik does not work either, because similar to entail, [the skbs pushed when the window is available](https://github.com/tempesta-tech/linux-5.10.35-tfw/blob/b463f54287d8865df5a48dd5961a0ce5c429145b/include/net/tcp.h#L1907) does not mean that all skbs can be sent immediately, because it does not know how large the network capacity is and how many skbs can be sent, so some skbs will be hold in the memory for an undetermined time. Again, the only sign is `tcp_transmit_skb`.

We cannot discard correctness for the sake of code simplicity, right?

Reply:
We call `ss_skb_on_tcp_entail` just before we push skb to socket write queue. For HEADERS and DATA frames we calculate tcp window in `sk_fill_write_queue` and don't call `ss_skb_tcp_entail`->`ss_skb_on_tcp_entail` if we have no enough window. To solve problem with control frames we should not call `ss_skb_tcp_entail`->`ss_skb_on_tcp_entail`  for all other type of frames if there is no enough tcp window. We should push them to the queue of postponed frames and push them to socket write queue only when we have enough TCP window. In this case we can counted control frames in `ss_skb_on_tcp_entail`

Reply:
Fair enough. This is sufficient if we do not care about the strict accuracy of whether the skb was successfully sent by `tcp_transimit_skb` or was even eliminated by `tcp_rtx_queue`.

Reply:
I agree that the `flags` are replaced with the `on_send` callback in #1973, so it's not desired to keep flags, especially with only one user. Meantine, I also don't like the idea of additional queueing since to wasted memory decreasing performance and opens opportunities for DoS attacks.

However, in first place, I didn't get what do we fix with #2108, so I requested changes. Maybe I'm wrong and let's discuss the fix, but I proposed to control and limit  flows to client in https://github.com/tempesta-tech/tempesta/issues/498#issuecomment-2188757361